부산대 BE_배근호 2단계 - 회원 가입 및 로그인 구현

hopencode opened the pull request:
회원가입 및 로그인 기능 구현했습니다.
페이지 쪽은 로그인 시 토큰을 유지한 채 정보 조회 페이지로 넘어가는 부분에서 어려움이 있어 아직 구현 중입니다.
페이지 구현까지 완료 후 시간이 남으면 비밀번호 변경이나 계정 삭제 등도 추가해볼 예정입니다.

hopencode committed:
docs : STEP2 README 작성

hopencode committed:
style : STEP2 README 줄바꿈 수정

hopencode committed:
feat: 회원 도메인 생성

hopencode committed:
fix: Member 엔티티에 id 속성 추가

hopencode committed:
feat: member 테이블 및 repository 레이어 기반 구현

hopencode committed:
feat: findByEmail 기능 구현

hopencode committed:
feat: 회원가입 기능 구현
    - MemberRepository save 구현
    - MemberService 인터페이스 및 서비스 구현
    - JwtAuth에서 토큰 생성 구현
    - Member 컨트롤러 구현
    - STEP2에서 발생 가능한 예외 MemberExceptions에 정의
    - 회원가입 시 이미 등록된 이메일 예외 핸들러에 추가

hopencode committed:
feat: 회원가입 정상 및 이미 등록된 이메일 테스트 코드 작성
    - 서비스 및 컨트롤러 반환 형식 수정

hopencode committed:
feat: 로그인 기능 구현 및 추후 토큰 유효성 검사 예외 정의

hopencode committed:
feat: 로그인 테스트 코드 작성

hopencode committed:
fix: RequestDto의 setter들 구현
    - setter가 없으면 thymleaf에서 form에 입력한 데이터를 DTO에 바인딩
      할 때 주입이 되지 않아 null값이 들어가서 오류가 발생함

hopencode committed:
feat: 회원가입 페이지 구현

hjjae2 commented on src/main/java/gift/auth/JwtAuth.java:
```diff
@@ -0,0 +1,21 @@
+
+@Component
+public class JwtAuth {
+    private String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
저장소에 올라와도 되는 값일까요? 숨겨야 하는 값으로 보여저서요. 🥲

hopencode replied:
secure.properties 파일에 키값을 저장하고 값을 불러와 사용하도록 수정했습니다.
secure.properties 파일은 저장소에 올라가지 않도록 gitignore에 추가했습니다.

hjjae2 commented on src/main/java/gift/auth/JwtAuth.java:
```diff
@@ -0,0 +1,21 @@
+        String accessToken = Jwts.builder()
+                .setSubject(member.getEmail())
+                .claim("email", member.getEmail())
+                .claim("password", member.getPassword())
```
password 정보를 추가해주시는 이유가 궁금합니다.

hopencode replied:
아 이건 초기 nextstep에 있는 예제 코드를 그대로 복사해서 기존 claim이 2개라 email과 password를 추가했다가 보안상 우려가 있을 듯 해 이후 커밋에서 password 부분은 제거했습니다.

hjjae2 commented on src/main/java/gift/controller/GlobalExceptionHandler.java:
에러 종류 별로 처리를 나누고 적절한 상태 코드를 잘 응답해주고 있네요. 꼼꼼한 처리 좋습니다. 👍

hjjae2 commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,35 @@
+        this.memberService = memberService;
+    }
+
+    @PostMapping("/register")
```
/register 라는 동사가 꼭 필요할까요?

잘 아시겠지만, RESEful API 에선 (될 수 있으면)명사 기반으로 작성하도록 권장하고 있습니다. ([참고](https://learn.microsoft.com/ko-kr/azure/architecture/best-practices/api-design#define-restful-web-api-resource-uris))

hopencode replied:
/membership으로 수정했습니다

hjjae2 replied:
`POST /api/members` 로 충분하지 않을까요-?

hopencode replied:
이번에 수정하진 않았는데 다음 리뷰 때 수정해두겠습니다.

hjjae2 commented on src/main/java/gift/dto/ProductRequestDto.java:
```diff
@@ -24,8 +24,11 @@ public ProductRequestDto(String name, Integer price, String imageUrl) {
 
 
     public String getName() { return name; }
+    public void setName(String name) { this.name = name; }
 
     public Integer getPrice() { return price; }
+    public void setPrice(Integer price) { this.price = price; }
 
     public String getImageUrl() { return imageUrl; }
+    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }
```
이 코드들이 추가된 이유가 궁금합니다. 🙇

hopencode replied:
생성자를 통해 테스트 및 내부 로직에서는 setter 없이 동작할 수 있도록 하였는데
이렇게 수정 후 setter를 지웠더니 상품 추가/수정 페이지에서 상품 정보 입력 후 요청을 보내면 에러가 발생했습니다.
thymleaf의 form에서 입력한 데이터를 dto에 주입할 때는 setter를 사용해서 없으면 에러가 나는 것 같아 추가했습니다.

hjjae2 replied:
> thymleaf의 form에서 입력한 데이터를 dto에 주입할 때는 setter를 사용해서 없으면 에러가 나는 것 같아 추가했습니다.

(변경을 요청드리는 것은 아닙니다.) 혹시 setter 없이 페이지를 구성하는 방법은 없을까요? 번거로우시겠지만 한번 확인해봐주시겠어요? 🙏 🙇

hopencode replied:
STEP3 PR에 setter 없이 생성자에 넣도록 수정했습니다

hjjae2 commented on src/main/java/gift/exception/MemberExceptions.java:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    public static class InvalidAuthorizationHeaderException extends RuntimeException {
+        public InvalidAuthorizationHeaderException() {
+            super("Authorization 헤더가 유효하지 않습니다.");
+        }
+    }
+
+    public static class InvalidTokenException extends RuntimeException {
+        public InvalidTokenException() {
+            super("토큰이 유효하지 않습니다.");
+        }
+    }
```
이 두 예외는 사용되는 곳이 없어보이는데, 이후 작업에서 사용될 예정일까요?

hopencode replied:
로그인 페이지에서 로그인 이후 관리자 페이지로 넘어갈 때 토큰을 검사하는 부분에서 사용하려고 미리 만들어두었습니다.

hjjae2 commented on src/main/java/gift/dto/MemberRequestDto.java:
```diff
@@ -0,0 +1,26 @@
+
+public class MemberRequestDto {
+    private String email;
+    private String password;
```
이전 단계에서 구현했던 것과 같이 값에 대한 검증을 추가해보시죠. 😄

hopencode replied:
형식이나 길이 제한 등 검증과 오류 메시지 반환 추가했습니다.

hjjae2 commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,45 @@
+        Member member = memberRepository.findByEmailAndPassword(requestDto.getEmail(), requestDto.getPassword())
+                .orElseThrow(MemberExceptions.InvalidPasswordException::new);
+
+        String token = jwtAuth.createJwtToken(member);
```
JWT 외 다른 방식으로 구현한다면 어떤 방식으로 인증(로그인 기능)을 구현해볼 수 있을까요?

hopencode replied:
클라이언트가 요청을 보낼 때 Authorization 헤더에 ID, 비밀번호 정보로 인코딩된 정보를 포함하여 서버에 요청함으로써 인증하는 방식이나
서버가 클라이언트로부터 인증 요청을 받으면 세션이라는 공간에 인식된 사용자의 정보를 저장하고 요청 헤더에 식별값을 추가해 전송하면 세션에서 식별값에 해당하는 정보를 찾아 로그인을 유지하는 세션 방식으로 구현할 수 있습니다.

hjjae2 commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Map;
+
+@RestController
+@RequestMapping("/api/members")
```
강의 시간에 잠깐 배우셨을거에요. 이렇게 작성해주신 컨트롤러는 `DispatcherServlet` 라는 녀석의 도움을 받아 사용자의 요청과 매핑되어 수행되게 되는데요. `DispatcherServlet` 에 대해 간략히 설명을 부탁드려도 될까요? 🙏

hopencode replied:
프론트 컨트롤러 역할로 
1. 들어오는 모든 요청을 받아 해당 요청을 처리하기에 적절한 컨트롤러를 찾아 전달하고 
2. 그렇게 전달하여 처리된 결과를 받아 어떤 뷰가 렌더링 할 지 결정하여 처리 결과를 전달한 뒤 
3. 해당 뷰로부터 받은 응답을 최종적으로 클라이언트에게 반환하는 것으로 알고 있습니다.

hjjae2 commented on src/main/java/gift/auth/JwtAuth.java:
```diff
@@ -0,0 +1,21 @@
+                .claim("password", member.getPassword())
+                .signWith(Keys.hmacShaKeyFor(secretKey.getBytes()))
+                .compact();
+        return accessToken;
```
JWT 를 사용하시네요 😄  JWT 가 무엇인지 궁금합니다, JWT 의 특징(형식, 장/단점)에 대해 간략히 설명해주실 수 있을까요?

hopencode replied:
JWT는  JSON 형식으 인증에 필요한 정보들을 토큰에 담아 암호화시켜 사용하는 방식으로 주로 로그인 후 사용자 인증 상태를 유지하는 데 사용됩니다.
JWT는 . (점) 을 기준으로 나뉘어 Header, Payload, Signature 세 가지로 구성됩니다.
여기서 Payload 부분에 ID, 권한 등 사용자 정보(Claims)가 담깁니다.
<br>

장점
- 로컬에 저장 -> 서버 용량에 영향 X
- 서버는 세션 저장 불필요, JWT 검증만→ 무상태(stateless) 유지 가능
- JSON 포맷 -> 확장성 우수
<br>

단점
- 쿠키/세션보다 토큰의 길이가 길어 → 인증 요청이 많아질수록 네트워크 부하
- Payload는 인코딩(B64)일 뿐 암호화가 아님 → 탈취 시 정보 노출 가능성 (보안 위험)

hjjae2 reviewed (changes requested):
안녕하세요, 근호님! 이번 단계도 꼼꼼하게 잘 구현해주셨네요, 고생하셨습니다. 👍

몇 가지 코멘트 남겼는데요, 확인 부탁드릴게요. 💪

hopencode committed:
refactor: RESTful API 권장에 따라 url 수정 (register -> membership)

hopencode committed:
refactor: jwt 토큰 생성을 위한 secretKey 숨기기
    - secure 프로퍼티를 만들어 거기에 secretKey 값을 숨기고
    - 해당 프로퍼티 파일은 저장소에 올리지 않도록 gitignore에 추가

hopencode committed:
feat: email 및 password 입력값 검증 추가

hjjae2 reviewed (approved):
안녕하세요, 근호님!

코멘트 드린 부분에 대해 상세히 잘 확인해주셔서 감사합니다. 🙇 잘 확인해주신 덕에 이번 단계는 여기서 마무리해도 좋겠습니다. 머지하고 다음 단계에서 뵙겠습니다.

이와 별개로 추가로 궁금한 내용을 코멘트 남겨두었는데요. 이것들은 확인 부탁드릴게요. 🙏
